File: UVA/On Try/495_Fibonacci freeze.cpp
#include<iostream.h>
long long dp[1000000];


long long fibonacci(long long n)
{

    long long &ans=dp[n];

    if(dp[n]!=0) return ans;

    if(n<=2)
    {
        return 1;

    }
    else
    {
        ans=fibonacci(n-1)+fibonacci(n-2);

        return ans;
    }
}

main()
{
    long long n;
    long long ans;
    ans=fibonacci(5000);

    while(scanf("%lld", &n)==1)
    {
        ans=fibonacci(n);
        printf("The Fibonacci number for %lld is %lld\n", n, ans);


    }
}

File: UVA/01. Accepted/08. Big Integer Problems/495_Fibonacci Freeze.java


import java.util.Scanner;
import java.math.BigInteger;

public class Main 
{
    public static void main(String[] args) 
    {
 		Scanner input = new Scanner(System.in);
 		BigInteger  ans[] = new BigInteger[5010];
 		int i;
 		BigInteger a, b;
 		a = new BigInteger("0");
 		b = new BigInteger("1");
 		
 		ans[0]=new BigInteger("1");
 		
 		for(i=1;i<=5000;i++)
 		{
 			ans[i] = a.add(b);
 			a = b;
 			b = ans[i];
 			
 		}
 		
 		int n;
 		while(input.hasNext())
 		{
 			n = input.nextInt();
 			i = n-1;
 			if(n==0)
 				System.out.println("The Fibonacci number for "+n+" is 0");
 			else
 				System.out.println("The Fibonacci number for "+n+" is "+ans[i]);
 			
 		}
      
    }
}
